Adding tests to prove expected behavior around force starting a quest

Hus274 opened the pull request:
part of #7653 
### Changes

Added tests to prove the following occurs as it is the expected behavior:
- check that the code for force-starting a quest removes any players from quest.members that don't have a true value
- check that the code for force-starting a quest removes any players from quest.members that aren't in the party

---

UUID: 2a4d30ee-4833-420e-be7c-ad6a6e261c5f

Hus274 committed:
starting

Hus274 committed:
adding tests to prove the expected behavior occurs around force starting a quest. part of #7653

crookedneighbor commented on test/api/v3/integration/quests/POST-groups_groupId_quests_force-start.test.js:
```diff
@@ -149,5 +149,64 @@ describe('POST /groups/:groupId/quests/force-start', () => {
+      await leader.post(`/groups/${questingGroup._id}/quests/invite/${PET_QUEST}`);
+
+      let notInPartyUser = await generateUser();
+      questingGroup.quest.members[notInPartyUser._id] = true;
```
This is only changing the local variable in the test. It's not adding the id to the object in the db.

crookedneighbor replied:
I want to say you should be able to do

``` js
await questingGroup.update({
  [`quest.members.${notInPartyUset._id}`]: true,
});
```

crookedneighbor replied:
^ does a direct db update

Hus274 replied:
I was trying to do this earlier but couldn't get it to work and then wasn't sure I needed it. It seems like I was missing the array brackets [ ] around the quest members thing in the update. Why are those needed? I thought you just needed the string defining the path

crookedneighbor commented on test/api/v3/integration/quests/POST-groups_groupId_quests_force-start.test.js:
```diff
@@ -149,5 +149,64 @@ describe('POST /groups/:groupId/quests/force-start', () => {
+      await leader.post(`/groups/${questingGroup._id}/quests/invite/${PET_QUEST}`);
+      await partyMembers[0].post(`/groups/${questingGroup._id}/quests/accept`);
+
+      questingGroup.quest.members[partyMemberThatRejects._id] = false;
```
Same comment as above.

Hus274 committed:
fixing up tests and adding functionality to remove user from quest if they are not in the questing party

Hus274 commented:
Fixed up the tests so they actually talk to the database correctly and added functionality to remove any users who aren't in the party from the party's quest.

crookedneighbor commented on website/server/models/group.js:
```diff
@@ -366,6 +366,17 @@ schema.methods.startQuest = async function startQuest (user) {
+  let partyId = this._id;
+  let questMembers = this.quest.members;
+  await Bluebird.map(Object.keys(this.quest.members), async (memberId) => {
+    let member = await User.findById(memberId, 'party._id').exec();
```
I'd do: 

``` js
let member = await User.findOne({_id: memberId, 'party.id': partyId});

if (!member) {
  delete questMembers[memberId];
}
```

This will catch users that don't exist.

crookedneighbor replied:
In fact, let's add a test for that case. The case where a user deleted their account and it wasn't cleaned up properly.

Hus274 committed:
adding code to handle when the user is deleted and added a test around that situation

Hus274 commented:
added code to handle the deleted user and a test around that case.

Hus274 committed:
Merge remote-tracking branch 'upstream/develop' into 7653-2

Hus274 commented:
the merge fixed the tests.

paglias commented on website/server/models/group.js:
```diff
@@ -366,6 +366,17 @@ schema.methods.startQuest = async function startQuest (user) {
+  let partyId = this._id;
+  let questMembers = this.quest.members;
+  await Bluebird.map(Object.keys(this.quest.members), async (memberId) => {
+    let member = await User.findOne({_id: memberId, 'party._id': partyId});
```
this way the entire user will be fetched. I'd do `User.findOne({_id: memberId, 'party._id': partyId}).select('_id').lean()` so only the _id will be returned and using `.lean()` will prevent a mongoose object from being build instead returning a plain js object which is much faster

Hus274 replied:
Makes sense to me, I'll update this and resubmit

khipkin commented on test/api/v3/integration/quests/POST-groups_groupId_quests_force-start.test.js:
```diff
@@ -149,5 +149,86 @@ describe('POST /groups/:groupId/quests/force-start', () => {
+      let partyMemberThatIgnores = partyMembers[2];
+
+      await leader.post(`/groups/${questingGroup._id}/quests/invite/${PET_QUEST}`);
+      await partyMembers[0].post(`/groups/${questingGroup._id}/quests/accept`);
```
`partyMemberThatAccepts` for consistency and clarity

khipkin commented on test/api/v3/integration/quests/POST-groups_groupId_quests_force-start.test.js:
```diff
@@ -149,5 +149,86 @@ describe('POST /groups/:groupId/quests/force-start', () => {
+
+      await leader.post(`/groups/${questingGroup._id}/quests/force-start`);
+
+      await Bluebird.delay(500);
```
Why is this necessary?

Hus274 replied:
If you look at the force start code it's not all synchronous, it kicks things off purposefully without waiting for them. If you look at the other tests in this file, the majority of them also have this.

khipkin commented on test/api/v3/integration/quests/POST-groups_groupId_quests_force-start.test.js:
```diff
@@ -149,5 +149,86 @@ describe('POST /groups/:groupId/quests/force-start', () => {
+      await questingGroup.update({
+        [`quest.members.${notInPartyUser._id}`]: true,
+      });
+      questingGroup.sync();
```
await?

Hus274 committed:
fixing up comments

Hus274 commented:
addressed comments

crookedneighbor commented on website/server/models/group.js:
```diff
@@ -366,6 +366,17 @@ schema.methods.startQuest = async function startQuest (user) {
+  // remove any users from quest.members who aren't in the party
+  let partyId = this._id;
+  let questMembers = this.quest.members;
+  await Bluebird.map(Object.keys(this.quest.members), async (memberId) => {
```
Since you've created the `questMembers` variable, we should probably use it here.

crookedneighbor replied:
Never mind, it's fine.

crookedneighbor commented:
Notated toward your next tier.
